### PR TITLE
Reduce spammer motivation (MBS-9353, MBS-9354 MBS-9355)

### PR DESCRIPTION
**Pull request by yvanzo**
- [MBS-9353](https://tickets.metabrainz.org/browse/MBS-9353): Do not show collections or user profiles to people not logged in
- <s>[MBS-9354](https://tickets.metabrainz.org/browse/MBS-9354): Block `/collection` as per `robots.txt`</s> (see https://github.com/metabrainz/musicbrainz-server/pull/519#discussion_r120005672 instead)
- [MBS-9355](https://tickets.metabrainz.org/browse/MBS-9355): Do not allow limited user to set bio/link

### CODE REVIEW

**Comment by yvanzo** [comment]
I did not update tests for now, `Controller` tests are currently blocked by silly rendering error in my dev env.

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm [hunk 1]
```diff
@@ -96,13 +96,14 @@ sub validate_birth_date {
     return $field->add_error("invalid date") unless Date::Calc::check_date($year, $month, $day);
 }
 
-sub _requires_email {
+sub _requires_not_limited {
     my ($self, $field) = @_;
-    return $field->add_error(l('Biography and website fields may not be edited until your email address is confirmed.')) unless $self->ctx->user->has_confirmed_email_address;
+    return $field->add_error(l('Spammers are not welcome, go away.')) unless $self->ctx->user->is_limited;
+    ;
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: There's a lone semi-colon there.

**Comment by mwiencek** [comment]
Looks good to me other than the broken tests. I'll approve the code, though I dislike hiding entities behind a log-in.

**Comment by mwiencek** [comment]
Are you still blocked by the rendering error? I can't trigger that anymore so I'm not sure how to fix it yet. :/ But I can look at fixing the tests if you want.

**Comment by yvanzo** [comment]
@brainzbot retest this please

**Comment by yvanzo** [comment]
Well, the rendering error is back and I did not succeed in bisecting current `master` branch.

> Catalyst error screen: Caught exception in MusicBrainz::Server::View::Default->process "Couldn't connect to the renderer. at /musicbrainz-server/lib/MusicBrainz/Server/View/Base.pm line 36."

Despite this, I updated tests for `User::Edit` but there seems to be a permanent error in `t/script/NotifyOfNewReleases.t` which might be unrelated to this patchset.

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm (hunk 1 above)
Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: Don't forget the wild ;

File: root/robots.txt.production [hunk 2]
```diff
@@ -5,6 +5,7 @@ Disallow: /ajax/
 Disallow: /area/
 Disallow: /cdstub/
 Disallow: /cdtoc/attach
+Disallow: /collection/
```
Comment (mwiencek) [inline root/robots.txt.production]: This commit will be removed automatically if you rebase.

File: t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm [hunk 3]
```diff
@@ -57,13 +51,76 @@ $mech->get_ok($verify_email_path);
...
+$mech->get_ok('/account/edit');
+my $tx = test_xpath_html($mech->content);
+html_ok($tx);
+$tx->ok(selector_to_xpath('input#"id-profile.email"'), 'email field for all users');
+$tx->not_ok(selector_to_xpath('input#"id-profile.website"'), 'no website field for limited users');
+$tx->not_ok(selector_to_xpath('input#"id-profile.biography"'), 'no biography field for limited users');
+
+$test->c->sql->do(<<EOSQL, $editor->id);
+    INSERT INTO edit (id, editor, type, status, expire_time, autoedit) VALUES
+        ( 1, 1, 1, $STATUS_APPLIED, now(), 0),
```
Comment (mwiencek) [inline t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm]: `$STATUS_APPLIED` needs to be imported.

File: t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm [hunk 4]
```diff
@@ -57,13 +51,76 @@ $mech->get_ok($verify_email_path);
...
+        ( 5, 1, 1, $STATUS_APPLIED, now(), 0),
+        ( 6, 1, 1, $STATUS_APPLIED, now(), 0),
+        ( 7, 1, 1, $STATUS_APPLIED, now(), 0),
+        ( 8, 1, 1, $STATUS_APPLIED, now(), 0),
+        ( 9, 1, 1, $STATUS_APPLIED, now(), 0),
+        (10, 1, 1, $STATUS_APPLIED, now(), 0);
+EOSQL
+
+$mech->get_ok('/account/edit');
+my $tx = test_xpath_html($mech->content);
```
Comment (mwiencek) [inline t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm]: `s/my //` since it's already declared above.

File: t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm [hunk 5]
```diff
@@ -57,13 +51,76 @@ $mech->get_ok($verify_email_path);
...
 
 $mech->get('/user/new_editor');
-$mech->content_contains('http://example.com/~new_editor/');
-$mech->content_contains('hello world!');
 $mech->content_contains('new_email@example.com');
 
 
 };
 
+test 'Limited users cannot edit website and biography' => sub {
```
Comment (mwiencek) [inline t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm]: Do we need the URL format/date checking tests in there too? They seem to duplicate what's above and are unrelated to the description, but I could be missing something.

File: t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm [hunk 6]
```diff
@@ -57,13 +51,76 @@ $mech->get_ok($verify_email_path);
...
+$mech->submit_form( with_fields => { username => 'new_editor', password => 'password' } );
+
+$mech->get_ok('/account/edit');
+my $tx = test_xpath_html($mech->content);
+html_ok($tx);
+$tx->ok(selector_to_xpath('input#"id-profile.email"'), 'email field for all users');
+$tx->not_ok(selector_to_xpath('input#"id-profile.website"'), 'no website field for limited users');
+$tx->not_ok(selector_to_xpath('input#"id-profile.biography"'), 'no biography field for limited users');
+
+$test->c->sql->do(<<EOSQL, $editor->id);
```
Comment (mwiencek) [inline t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm]: There's no `$editor` variable declared.

**Comment by mwiencek** [comment]
Sorry for the template renderer issues, I'm still stumped by that.

**Comment by yvanzo** [comment]
@brainzbot retest this please

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm (hunk 1 above)
Comment (yvanzo) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: Fixed in https://github.com/metabrainz/musicbrainz-server/pull/515/commits/76c82854c6558e8dfac0100266fa31926db59c59

File: root/robots.txt.production (hunk 2 above)
Comment (yvanzo) [inline root/robots.txt.production]: I will rebase once everything else gets fixed/accepted.

File: t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm (hunk 5 above)
Comment (yvanzo) [inline t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm]: URL format yes, date checking no. Fixed in https://github.com/metabrainz/musicbrainz-server/pull/515/commits/cf46dea1e27697d92d3ea9201419c5c0f2bb165c

File: t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm (hunk 6 above)
Comment (yvanzo) [inline t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm]: Fixed in https://github.com/metabrainz/musicbrainz-server/pull/515/commits/cf46dea1e27697d92d3ea9201419c5c0f2bb165c

File: t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm (hunk 3 above)
Comment (yvanzo) [inline t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm]: Fixed in https://github.com/metabrainz/musicbrainz-server/pull/515/commits/cf46dea1e27697d92d3ea9201419c5c0f2bb165c

File: t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm (hunk 4 above)
Comment (yvanzo) [inline t/lib/t/MusicBrainz/Server/Controller/User/Edit.pm]: Fixed in https://github.com/metabrainz/musicbrainz-server/pull/515/commits/cf46dea1e27697d92d3ea9201419c5c0f2bb165c

**Comment by yvanzo** [comment]
Thanks for your review, I fixed every single point I couldn't find without being able to run tests.

I don't know why tests are failing on Jenkins [Build #439](https://ci.metabrainz.org/job/musicbrainz-server/439/) (reproduced with [Build #440](https://ci.metabrainz.org/job/musicbrainz-server/440/)) whereas they pass locally. These HTML/encoding errors are weird since templates were just slightly modified.

I will make a proper rebase once everything gets fixed.

**Comment by mwiencek** [comment]
Well, I can fix the HTML validation errors with https://gist.github.com/mwiencek/b0f383ff4474f3165dbdfe5d63735715

It's possible those were skipped for you locally because you don't have an HTML validator configured. But even after fixing these, I still get the xpath failures ("not ok 4 - email field for all users" etc.)—those are passing for you too?

**Comment by mwiencek** [comment]
BTW, I have an HTML validator service running locally with docker, if this helps:

```bash
docker run \
    --detach \
    --name html5-validator \
    --publish 8888:8888 \
    magnetikonline/html5validator
```

```perl
# lib/DBDefs.pm
sub HTML_VALIDATOR { 'http://localhost:8888/?out=json' }
```

**Comment by yvanzo** [comment]
Thanks, I forgot about HTML validator. It looks like `HTML::Selector::XPath` may not support dot in element ID. I got rid of it and fixed careless mistakes in both tests and actual code.

File: lib/MusicBrainz/Server/Form/User/EditProfile.pm [hunk 7]
```diff
@@ -96,13 +96,13 @@ sub validate_birth_date {
     return $field->add_error("invalid date") unless Date::Calc::check_date($year, $month, $day);
 }
 
-sub _requires_email {
+sub _requires_not_limited {
     my ($self, $field) = @_;
-    return $field->add_error(l('Biography and website fields may not be edited until your email address is confirmed.')) unless $self->ctx->user->has_confirmed_email_address;
+    return $field->add_error(l('Spammers are not welcome, go away.')) unless !$self->ctx->user->is_limited;
```
Comment (mwiencek) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: Two small things: (1) I think we don't have to translate this message, since it only shows up if they're abusing the form (the relevant fields are hidden), and (2) `unless !` hurt my head, I think an `if` would suffice. :)

Comment (yvanzo) [inline lib/MusicBrainz/Server/Form/User/EditProfile.pm]: I don’t disagree here 😉 Fixed in https://github.com/metabrainz/musicbrainz-server/pull/515/commits/4f3df78c7182cd7343a2e13d59bb0edd5dcc80cf.

**Comment by yvanzo** [comment]
Rebased.
